pipeline/scripts/ci/upload_compose.py

```diff
@@ -0,0 +1,151 @@
+"""
+This script executes the development build upload workflow to IBM COS.
+
+The steps followed are
+- Create the necessary repo files
+- Perform a repo sync on all required repositories
+- Compress the repositories and upload them to COS
+
+Note:
+    The script is required to be executed with sudo privileges. This is required as
+    we would be creating repo files for sync operations.
+"""
+import logging
+import random
+import string
+import subprocess
+import sys
+import tempfile
+from os.path import join
+
+from docopt import docopt
+from jinja2 import Template
+
+from storage.ibm_cos import CloudObjectStorage
+
+LOG = logging.getLogger(__name__)
+REPO_TEMPLATE = """
+{%- for repo in ["OSD", "MON", "Tools"] -%}
+[{{ data.nvr }}-{{ repo }}]
+name = {{ data.nvr }}-{{ repo}}
+baseurl = {{ data.base_url }}/compose/{{ repo }}/$basearch/os
+enabled = 0
+gpgcheck = 0
+
+{% endfor %}
+"""
+
+usage = """Upload Compose.
+
+This script pulls the development RPMs available in the given base_url and uploads an
+archive of it using the provided object_name to the mentioned bucket.
+
+In the context of CephCI,
+    bucket_name     is the name of the RHCS build
+    object_name     is the ceph version
+    base_url        the base url containing the RPMs to be uploaded
+
+  Example:
+    python upload_compose.py rhceph-4.2-rhel-7 ceph-14.6.2-113 http://download.rh.com/x
+
+Usage:
+  upload_compose.py <bucket_name> <object_name> <base_url>
+  upload_compose.py -h | --help
+
+Options:
+  -h --help     Show this screen
+
+"""
+
+
+def create_repo(build: str, url: str) -> None:
+    """
+    Creates all the required repo files based on the given base_url.
+
+    Args:
+        build: str          Bucket name or RH Ceph build
+        url: str       URL holding the RPMs to be uploaded
+    Returns:
+        None
+    """
+    _tmpl = Template(REPO_TEMPLATE)
+    data = dict({"nvr": build, "base_url": url})
+    repo_file = _tmpl.render(data=data)
+    LOG.debug(f"The repo file is \n {repo_file}")
+
+    with open(f"/etc/yum.repos.d/{build}.repo", "w") as fh:
+        fh.write(repo_file)
+
+    LOG.info("Successfully created repo file.")
+
+
+def compress_build(build: str) -> str:
+    """
+    Returns the compressed file name containing the development RPMs.
+
+    Args:
+        build: str  The base repo id to be used to pull the packages.
+    Returns:
+        str: The complete file path to the archive file created.
+    """
+    temp_dir = tempfile.mkdtemp()
+    _file = "".join(random.choices(string.ascii_lowercase + string.digits, k=5))
+    temp_file = join("/tmp", f"{_file}.tar.gz")
+
+    for repo in ["OSD", "MON", "Tools"]:
+        subprocess.run(
+            [
+                "reposync",
+                "-l",
+                "-d",
+                "-m",
+                "--newest-only",
+                "--download-metadata",
+                "--repoid",
+                f"{build}-{repo}",
```

Comment:
You can define repos dynamically on the CLI here without having to write them into `/etc/` beforehand. The CLI flag is `--repofrompath`.

Here's an example that dynamically defines a repo named "foobar", and then syncs all the content from that:

```
reposync \
  -l \
  -m \
  --newest-only \
  --download-metadata \
  --repoid foobar \
  --repofrompath=foobar,https://download.example.com/ceph/x86_64/ \
  --download-path /tmp/brew
```

I think that will allow you to remove the `create_repo()` method

Reply:
`--repofrompath` is unavailable in `centos-7` hence we do need the `create_repo()` method. During our earlier testing, we couldn't use `rhel-8` as we had issues in sending emails from `cephci`.

We have disabled sending emails from `cephci`... let me add a user story/task to have us migrate the Jenkins node to `rhel-8`

Reply:
Once we have moved to `rhel-8`, I would be able to address/use `repofrompath`. I does simplify the workflow. Thanks for the feedback.